utils: Add uuid_xor_to_uint32 helper

asias opened the pull request:
Convert the uuid to a shard_id which is 32 bits. It is useful to choose a shard from a uuid.

Refs: #16927

bhalevy commented:
I think this belongs to the app layer, not to the library (util).  One can get the std::hash of the uuid as an integral number and calculate a modulo over smp::count.

Also, we bake the shard id into time uuids for uniqueness, and having this proposed helper would confusing, since I'd expect it to extract the shard id from the uuid value in the time uuid case.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6188/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6188/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7489/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6188/)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider1.cloudius-systems.com

asias commented:
> I think this belongs to the app layer, not to the library (util). One can get the std::hash of the uuid as an integral number and calculate a modulo over smp::count.
> 
> Also, we bake the shard id into time uuids for uniqueness, and having this proposed helper would confusing, since I'd expect it to extract the shard id from the uuid value in the time uuid case.

I agree it is not a good idea to put inside the UUID class.  I moved it out and renamed it.

asias commented:
> I think this belongs to the app layer, not to the library (util). One can get the std::hash of the uuid as an integral number and calculate a modulo over smp::count.

I did not use std::hash<utils::UUID> because I wanted to xor all the 4 32bits to a single 32 bits.  I took a second look.  Our hash function has a extra xor with the higher 32 bits into the lower 32 bits. So even if we only use the lower 32 bits of the hash<utils::UUID>, it has the entropy from all the bits. I will use the hash<utils::UUID> for the helper. 

```
namespace std {
template<>
struct hash<utils::UUID> {
    size_t operator()(const utils::UUID& id) const noexcept {
        auto hilo = id.get_most_significant_bits()
                ^ id.get_least_significant_bits();
        return size_t((hilo >> 32) ^ hilo);
    }
};
```

> 
> Also, we bake the shard id into time uuids for uniqueness, and having this proposed helper would confusing, since I'd expect it to extract the shard id from the uuid value in the time uuid case.

asias commented on test/boost/UUID_test.cc:
```diff
@@ -100,6 +100,18 @@ BOOST_AUTO_TEST_CASE(test_get_time_uuid) {
 }
 
+BOOST_AUTO_TEST_CASE(test_uuid_to_uint32) {
+    using namespace std::chrono;
```
It is copied by accident. I will drop.

tchaikov commented:
> #### 🔴 CI State: FAILURE
> 
>    * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 
>    * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 

looks like #17029

>    * [topology_experimental_raft.test_tablets.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.6)
> 
>    * [topology_experimental_raft.test_tablets.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.6)
> 
> 
> #### Build Details:
> 
>     * Duration: 3 hr 41 min
> 
>     * Builder: i-0248786d2f4ec216a (m5ad.12xlarge)

denesb commented:
> #### Failed Tests (4/24109):
> 
>     * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 
>     * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 
>     * [topology_experimental_raft.test_tablets.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_debug_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.debug.6)
> 
>     * [topology_experimental_raft.test_tablets.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6193/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.6)

https://github.com/scylladb/scylladb/issues/17029

asias commented:
@denesb ping for merge.

denesb commented:
Queued.
